### ISSUE TITLE
EKS and AKS cluster waiting for API indefinitely

### ISSUE THREAD

**Issue by alexandrem** [rancher/rancher#15115]
**Rancher versions:**
rancher/server or rancher/rancher: v2.0.7
rancher/agent or rancher/rancher-agent: v2.0.7

**Type/provider of hosts: (VirtualBox/Bare-metal/AWS/GCE/DO)** AWS or Azure

**Steps to Reproduce:**

Deploy Rancher in Kubernetes in corporate environment (not externally available).

- Create a EKS cluster
- Resources get created successfully on the AWS account, cluster is healthy and active

**Results:**

Rancher cannot validate that cluster is up and available, it always says

```
This cluster is currently Waiting; areas that interact directly with it will not be available until the API is ready.

Waiting for API to be available
```

In the cattle server logs, I can see this:

```
2018/08/17 00:56:30 [ERROR] ClusterController c-t5pks [cluster-agent-controller] failed with : could not contact server: Get https://0D25CAAA4FD47B08DAB110FD20665432.sk1.us-west-2.eks.amazonaws.com/version?timeout=30s: proxyconnect tcp: waiting for cluster agent to connect
```

We have a corporate proxy and all environment variables are defined correctly.

I wonder if this part of the code has a problem with that.

It's also not clear to me if having the Rancher API available externally is mandatory for EKS/AWS. I don't see anything specific deployed in the workloads to reach back to us, so I doubt it.

I can create GKE clusters successfully without any issue and they appear as healthy in Rancher.

Been trying to intercept these http requests with tcpdump from the container and I'm not even sure that the requests go out. DNS query is made though.

**Comment by alexandrem** [comment]
One notable issue also is that the remote cluster is never correctly "linked" to Rancher cluster objects, so deleting the cluster in Rancher will not delete it in the remote EKS account.

**Comment by s1113950** [comment]
I'm running into the same thing (also behind corporate proxy). I set `HTTP_PROXY` and `HTTPS_PROXY` vars when doing a `docker run ... rancher/rancher:latest` and that enabled me to get past the `service account` file portion, but when I tried to create a new cluster I get this:
```
2018/08/17 20:57:22 [ERROR] ClusterController c-dww4x [cluster-agent-controller] failed with : could not contact server: Get https://XXXXXXXX/version?timeout=30s: proxyconnect tcp: wai
ting for cluster agent to connect
```

**Comment by guangbochen** [comment]
This more seems like a proxy issue with your server configuration or your browser, if you have for example `SwitchOmega` extension enabled in your Chrome, try to add the rancher server URL to the byPass List, if it is a rancher server proxy issue, can you please provide more info of your configuration, thanks.

**Comment by alexandrem** [comment]
It's not a proxy server issue in my corporate environment.

I tested with a separate squid instance with default config and it had the same behavior.

Also, it happens that `eks.amazonaws.com` is actually whitelisted in our network and can be reached without the proxy. So I added `.eks.amazonaws.com` in the NO_PROXY var and succeeded in having the cluster provisioned and monitored as healthy correctly by Rancher (I have to keep the other http_proxy env vars because some provisioning part like the stack and VPC need to go through the proxy).

Note that this is same behavior for AKS.

I believe there is something wrong in the part of the code that handles the http calls for the post checks.

I figure this is in the kontainer-engine code base in aks_driver.go and eks_driver.go

I'm pretty sure this could be reproduced easily with a local rancher instance via docker run and a local squid proxy.

**Comment by guangbochen** [comment]
well I see,  this is because your rancher server is behind the public network so the AKS or EKS's callback cannot connect to your rancher server successfully, you will need to add a proxy to expose your rancher server and make sure the proxy IP and port and configured properly within the rancher settings `https://your-rancher-server-url/g/settings/advanced`. 
![image](https://user-images.githubusercontent.com/4569037/44345324-9b8cab00-a4c5-11e8-827c-da9f76a27426.png)

**Comment by alexandrem** [comment]
@guangbochen 

I saw that architecture diagram and it doesn't seem accurate for hosted Kubernetes providers.

You don't have to make the rancher server URL be externally reachable to provision clusters in EKS and GKE (AKS need to be confirmed but probably same).

There's nothing to call back from that remote clusters, there's no RKE provisioning and no cattle agent being deployed.

It's easy to validate by launching your own rancher server locally via `docker run -p 80:80 -p 443:443 -v $(pwd)/var-lib-rancher:/var/lib/rancher --name rancher-server rancher/rancher:v2.0.7` behind NAT.

Your server is not exposed to the internet and you can deploy clusters to GKE/EKS successfully.

Once again, I can successfully deploy GKE/EKS clusters in my corporate environment without any internet ingress by skipping the proxy connect for the remote k8s apiserver checks.

**Comment by alexandrem** [comment]
Ok EKS and AKS  both have cattle cluster agent and node agents deployed.

It's definitely not good to not have the rancher server URL externally reachable for complete functionality, however this is past the part about the proxy connection issue.

At least we can see the cluster status and the list of nodes for your cluster when not using proxy, and validate the container workloads via Rancher UI. When on proxy, you can't do that at all, the cluster status page is stucked at "waiting for API..."

**Comment by sybnex** [comment]
We are facing the exact same issue. 
![image](https://user-images.githubusercontent.com/12811138/44906785-2cf5eb80-ad16-11e8-9a9e-b0ff6cea5fbf.png)

Rancher is launched with correct PROXY and NO_PROXY Stuff. We build a setup like on a AirGap Installation. 
Rancher is able to launch the AKS which builds everything including VPC connectivity. We have configured our firewall and the routes as well as the whole DNS stuff on the nodes. The Nodes are able to connect to rancher and e.g. the docker registry. 
But then it is just in this "**Waiting for API to be available**" state. And the in the logs it is like: "**[cluster-agent-controller] failed with : ClusterUnavailable 503: cluster not found**".

I connected with the AKS certifikates from outside of the whole environement to kubernetes directly with kubectl. So everthing is just ready to deploy but nothing happend.
